### ISSUE TITLE
Incorrect pre-LN in the transformer model

### ISSUE THREAD

**Issue by csukuangfj** [pytorch/pytorch#61165]
The recent pull request #60593 introduced `norm_first`, i.e., pre-LayerNorm, into transformer models. 

However, its implementation does NOT match the one proposed in various papers. 
Also, it differs from the one used in ESPnet. So I think its implementation is incorrect.

---

## Current implementation from #60593

https://github.com/pytorch/pytorch/blob/54673fc944110a615bf4eadb231bd4bb114ea22f/torch/nn/modules/transformer.py#L324-L332

The line
```
src = self.norm1(src)
```
is problematic as it does not save a copy of the input.

## The one used by ESPnet
See https://github.com/espnet/espnet/blob/master/espnet/nets/pytorch_backend/transformer/encoder_layer.py#L71
```python
        residual = x
        if self.normalize_before:
            x = self.norm1(x)
```

## The one from the paper "On Layer Normalization in the Transformer Architecture"

The following figure is from https://arxiv.org/pdf/2002.04745.pdf

<img width="498" alt="Screen Shot 2021-07-02 at 1 28 10 PM" src="https://user-images.githubusercontent.com/5284924/124225039-5cb64500-db39-11eb-8676-f7daba690730.png">

You can see from (b), we should save a copy of the input `x_l`

## The one from the paper "Transformers without Tears: Improving the Normalization of Self-Attention"

The equation shown below is from https://arxiv.org/pdf/1910.05895.pdf

<img width="347" alt="Screen Shot 2021-07-02 at 1 30 12 PM" src="https://user-images.githubusercontent.com/5284924/124225247-b0c12980-db39-11eb-9b64-85b28548afda.png">

---

@bhosmer 
@jbschlosser

cc @ezyang @gchanan @zou3519 @bdhirsh @jbschlosser @anjali411 @albanD @mruberry @bhosmer @cpuhrsch

**Comment by heitorschueroff** [comment]
Adding to triage review to discuss priority

**Comment by bhosmer** [comment]
I can take a look at this later in the week. The precedent cited here definitely seems compelling, just want to review the original issues to recall what prompted the current implementation

**Comment by jbschlosser** [comment]
Fixed in #61692
